compiler/src/iree/compiler/Codegen/Dialect/VectorExt/Transforms/VectorizeIREEVectorExtOps.cpp

```diff
@@ -121,4 +125,365 @@ struct VectorizeIREEVectorExtOpsPass final
+                                 map.getNumResults());
+          found = true;
+        } else {
+          // Allow 0-D and 1-D maps.
+          AffineMap indexMap =
+              genericOp.getIndexingMapsArray()[blockArg.getArgNumber()];
+          if (indexMap.getNumResults() == 0) {
+            map = map.insertResult(getAffineConstantExpr(0, ctx),
+                                   map.getNumResults());
+            found = true;
+          } else if (indexMap.getNumResults() == 1) {
+            if (auto dimExpr = dyn_cast<AffineDimExpr>(indexMap.getResult(0))) {
+              map =
+                  map.insertResult(getAffineDimExpr(dimExpr.getPosition(), ctx),
+                                   map.getNumResults());
+              found = true;
+            }
+          }
+        }
+        break;
+      }
+
+      // Trace back arith.index_cast ops. This is a commonly occuring case,
+      // there may be other cases which we can trace back here (For example,
+      // addition with a loop invariant constant).
+      if (isa<arith::IndexCastOp>(defOp)) {
+        trace = defOp->getOperand(0);
+      } else {
+        break;
+      }
+    }
+
+    if (!found) {
+      return AffineMap();
+    }
+  }
+
+  return map;
+}
+
+static AffineMap inversePerm(AffineMap map) {
+  MLIRContext *context = map.getContext();
+  AffineExpr zero = mlir::getAffineConstantExpr(0, context);
+  // Start with all the results as 0.
+  SmallVector<AffineExpr, 4> exprs(map.getNumInputs(), zero);
+  for (unsigned i : llvm::seq(unsigned(0), map.getNumResults())) {
+    // Skip zeros from input map. 'exprs' is already initialized to zero.
+    if (auto constExpr = dyn_cast<AffineConstantExpr>(map.getResult(i))) {
+      assert(constExpr.getValue() == 0 &&
+             "Unexpected constant in projected permutation");
+      (void)constExpr;
+      continue;
+    }
+
+    // Reverse each dimension existing in the original map result.
+    exprs[map.getDimPosition(i)] = getAffineDimExpr(i, context);
+  }
+  return AffineMap::get(map.getNumResults(), /*symbolCount=*/0, exprs, context);
+}
+
+} // namespace
+
+LogicalResult vectorizeGatherLikeGenericToTransferGather(
+    RewriterBase &rewriter, linalg::GenericOp linalgOp,
+    ArrayRef<int64_t> vectorSizes, ArrayRef<bool> scalableVecDims,
+    bool vectorizeNDExtract, bool flatten1DDepthwiseConv) {
+
+  if (!LinalgExt::isGatherlikeOp(linalgOp)) {
+    return linalg::vectorize(rewriter, linalgOp, vectorSizes, scalableVecDims,
+                             vectorizeNDExtract, flatten1DDepthwiseConv);
+  }
+
+  Location loc = linalgOp->getLoc();
+
+  SmallVector<int64_t> canonicalVectorSizes(vectorSizes);
+  SmallVector<bool> canonicalScalableDims(scalableVecDims);
+
+  // If vector sizes are not provided, assume static vector sizes and use loop
+  // ranges.
+  if (vectorSizes.empty()) {
+    canonicalVectorSizes = linalgOp.getStaticLoopRanges();
+    canonicalScalableDims.append(linalgOp.getNumLoops(), false);
```

Comment:
nit: assert or require that `canonicalScalableDims` is empty too since this is appending.